Implement Permissions API integration with the Storage Access API

charliewolfe opened the pull request:
#### 4164839ade95d9e06b82d50edb8adf491af98069
<pre>
Implement Permissions API integration with the Storage Access API
<a href="https://bugs.webkit.org/show_bug.cgi?id=297181">https://bugs.webkit.org/show_bug.cgi?id=297181</a>
<a href="https://rdar.apple.com/147948198">rdar://147948198</a>

Reviewed by BJ Burg.

This patch enables Permissions API query and onchange functionality to work with the Storage Access API.
More details below.

This progresses the following WPT, but it cannot be tested in CI since we don&apos;t have a local DNS resolver:
imported/w3c/web-platform-tests/storage-access-api/storage-access-permission.sub.https.window.html

<a href="https://privacycg.github.io/storage-access/#permissions-integration">https://privacycg.github.io/storage-access/#permissions-integration</a>

* LayoutTests/http/tests/permissions/resources/storage-access-permission-query-iframe.html: Added.
* LayoutTests/http/tests/permissions/storage-access-permissions-query-expected.txt: Added.
* LayoutTests/http/tests/permissions/storage-access-permissions-query.html: Added.
Add a layout test for coverage since the WPT will fail in CI.

* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-cross-origin-fetch.sub.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.BlobURLDedicatedWorker.sub.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.BlobURLSharedWorker.sub.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.thirdPartyBlobStorage.sub.https.window-expected.txt:

* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
(WebCore::MainThreadPermissionObserver::addChangeListener):
(WebCore::MainThreadPermissionObserver::removeChangeListener):
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::eventListenersDidChange):
Add or remove listeners from the network process when change event listeners are modified.

* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::isAllowedByPermissionsPolicy):
(WebCore::Permissions::toPermissionName):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::toFeatureNameForLogging):
(WebCore::readFeatureIdentifier):
(WebCore::defaultAllowlistValue):
* Source/WebCore/html/PermissionsPolicy.h:
Add permissions policy support for storage-access permission.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::queryStorageAccessPermission):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/StorageAccessPermissionChangeObserver.h:
(WebKit::StorageAccessPermissionChangeObserver::~StorageAccessPermissionChangeObserver):
Add an observer to listen for storage access permission changes for a web process, similar to
CookieChangeObserver.

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::queryStorageAccessPermission):
(WebKit::WebResourceLoadStatisticsStore::startListeningForStorageAccessPermissionChanges):
(WebKit::WebResourceLoadStatisticsStore::stopListeningForStorageAccessPermissionChanges):
Both stop listening functions are needed to remove listeners when the web process terminates or when
change listeners are removed.

(WebKit::WebResourceLoadStatisticsStore::wasGrantedStorageAccessPermissionInPage):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
Add a list to track observers for web processes listening to storage access permission changes.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess):
Remove the observer stored on WebResourceLoadStatisticsStore when a web process terminates.

(WebKit::NetworkConnectionToWebProcess::subscribeToStorageAccessPermissionChanges):
(WebKit::NetworkConnectionToWebProcess::unsubscribeFromStorageAccessPermissionChanges):
(WebKit::NetworkConnectionToWebProcess::storageAccessPermissionChanged):
(WebKit::NetworkConnectionToWebProcess::queryStorageAccessPermission):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::storageAccessPermissionChanged):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
Storage access permissions are controlled by the network process, not the client, so we need to add a new
IPC message to query permissions from the network process.

(WebKit::WebPermissionController::notifyObserversIfNeeded):
(WebKit::WebPermissionController::storageAccessPermissionChanged):
(WebKit::WebPermissionController::permissionChanged):
Storage access permissions are keyed by top-level site and subframe site, not top-level origin. So I
added storageAccessPermissionChanged for storage access permission changes and notifyObserversIfNeeded to
reduce code duplication.

(WebKit::WebPermissionController::addChangeListener):
(WebKit::WebPermissionController::removeChangeListener):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
Notify the network process when change listeners are added or removed so it only notifies web processes
of permission changes when listeners exist.

Canonical link: <a href="https://commits.webkit.org/298532@main">https://commits.webkit.org/298532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7c5c28f2a58db248054e60d5edfdae543b11cc8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115806 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121854 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c139c11c-1acf-4fed-96dc-93d985678c2d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117695 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44047 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/66328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dd3bd7c5-0b87-468a-bcca-f5101d167235) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118754 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103920 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68374 "Passed tests") | | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/27983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/22033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65525 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/98237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125005 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/42693 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100102 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96511 "Passed tests") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/41772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18507 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48172 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/42049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/45383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/43756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
```diff
@@ -38,6 +38,7 @@
 namespace WebCore {
 
 class Page;
+class RegistrableDomain;
```
Any reason not to use Site like the specification does? We should really stop using RegistrableDomain for most things.

charliewolfe replied:
The on-disk database that tracks granted storage access uses RegistrableDomain, so I’m matching that here. Since `permissions.query` directly references that database, I don’t think we can use Site in this change without migrating the database first.

burg commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -199,6 +199,11 @@ NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess()
     if (CheckedPtr networkStorageSession = storageSession())
         networkStorageSession->removeCookiesEnabledStateObserver(*this);
 
+    if (CheckedPtr networkSession = this->networkSession()) {
```
Do we prefer `this->networkSession()` or `networkSession()` ?

charliewolfe replied:
I think `networkSession()`, but `this->` is needed in this case because the variable name is the same as the function name.

charliewolfe committed:
Implement Permissions API integration with the Storage Access API
https://bugs.webkit.org/show_bug.cgi?id=297181
rdar://147948198

Reviewed by BJ Burg.

This patch enables Permissions API query and onchange functionality to work with the Storage Access API.
More details below.

This progresses the following WPT, but it cannot be tested in CI since we don't have a local DNS resolver:
imported/w3c/web-platform-tests/storage-access-api/storage-access-permission.sub.https.window.html

https://privacycg.github.io/storage-access/#permissions-integration

* LayoutTests/http/tests/permissions/resources/storage-access-permission-query-iframe.html: Added.
* LayoutTests/http/tests/permissions/storage-access-permissions-query-expected.txt: Added.
* LayoutTests/http/tests/permissions/storage-access-permissions-query.html: Added.
Add a layout test for coverage since the WPT will fail in CI.

* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/requestStorageAccess-cross-origin-fetch.sub.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.BlobURLDedicatedWorker.sub.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.BlobURLSharedWorker.sub.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage-access-api/storage-access-beyond-cookies.thirdPartyBlobStorage.sub.https.window-expected.txt:

* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
(WebCore::MainThreadPermissionObserver::addChangeListener):
(WebCore::MainThreadPermissionObserver::removeChangeListener):
* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::eventListenersDidChange):
Add or remove listeners from the network process when change event listeners are modified.

* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::isAllowedByPermissionsPolicy):
(WebCore::Permissions::toPermissionName):
* Source/WebCore/html/PermissionsPolicy.cpp:
(WebCore::toFeatureNameForLogging):
(WebCore::readFeatureIdentifier):
(WebCore::defaultAllowlistValue):
* Source/WebCore/html/PermissionsPolicy.h:
Add permissions policy support for storage-access permission.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::queryStorageAccessPermission):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/StorageAccessPermissionChangeObserver.h:
(WebKit::StorageAccessPermissionChangeObserver::~StorageAccessPermissionChangeObserver):
Add an observer to listen for storage access permission changes for a web process, similar to
CookieChangeObserver.

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::queryStorageAccessPermission):
(WebKit::WebResourceLoadStatisticsStore::startListeningForStorageAccessPermissionChanges):
(WebKit::WebResourceLoadStatisticsStore::stopListeningForStorageAccessPermissionChanges):
Both stop listening functions are needed to remove listeners when the web process terminates or when
change listeners are removed.

(WebKit::WebResourceLoadStatisticsStore::wasGrantedStorageAccessPermissionInPage):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
Add a list to track observers for web processes listening to storage access permission changes.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess):
Remove the observer stored on WebResourceLoadStatisticsStore when a web process terminates.

(WebKit::NetworkConnectionToWebProcess::subscribeToStorageAccessPermissionChanges):
(WebKit::NetworkConnectionToWebProcess::unsubscribeFromStorageAccessPermissionChanges):
(WebKit::NetworkConnectionToWebProcess::storageAccessPermissionChanged):
(WebKit::NetworkConnectionToWebProcess::queryStorageAccessPermission):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::storageAccessPermissionChanged):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
Storage access permissions are controlled by the network process, not the client, so we need to add a new
IPC message to query permissions from the network process.

(WebKit::WebPermissionController::notifyObserversIfNeeded):
(WebKit::WebPermissionController::storageAccessPermissionChanged):
(WebKit::WebPermissionController::permissionChanged):
Storage access permissions are keyed by top-level site and subframe site, not top-level origin. So I
added storageAccessPermissionChanged for storage access permission changes and notifyObserversIfNeeded to
reduce code duplication.

(WebKit::WebPermissionController::addChangeListener):
(WebKit::WebPermissionController::removeChangeListener):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
Notify the network process when change listeners are added or removed so it only notifies web processes
of permission changes when listeners exist.

Canonical link: https://commits.webkit.org/298532@main
